source/defclass-star.lisp

```diff
@@ -603,18 +612,31 @@ Example:
                                  forms))
          ;; NOTE: `set-difference' can shuffle the order of elements, thus `remove-if'.
          (method-body (remove-if (lambda (f) (member f options :test #'equal)) forms))
+         (export-generic-name-p (first (member :export-generic-name-p options :key #'first)))
+         (options (remove export-generic-name-p options :test #'equal))
+         (export-generic-name-p (second export-generic-name-p))
          (generalized-arglist (generalize-arglist arglist)))
     (when (and (not (equal generalized-arglist arglist))
                (every #'null (list method-body options)))
       (style-warn "Specialized arglist used without method body in define-generic: ~a" arglist))
-    `(defgeneric ,name ,generalized-arglist
-       ,@declarations
-       ,@(when method-body
-           `((:method ,arglist
-               ,@method-body)))
-       ,@options
-       ,@(when documentation
-           `((:documentation ,documentation))))))
+    `(progn
+       (defgeneric ,name ,generalized-arglist
+         ,@declarations
+         ,@(when method-body
+             `((:method ,arglist
+                 ,@method-body)))
+         ,@options
+         ,@(when documentation
+             `((:documentation ,documentation))))
+       ;; FIXME: Maybe check *PACKAGE*?
+       ,@(when (or export-generic-name-p *export-generic-name-p*)
+           `((eval-when (:compile-toplevel :load-toplevel :execute)
+               (export ',name ,(package-name *package*)))))
+       ,@(let ((documentation (or documentation
+                                  (second (first (member :documentation options :key #'first))))))
+           (when documentation
+             `((setf (documentation ',name 'function) ,documentation)
+               (setf (documentation (fdefinition ',name) 'function) ,documentation)))))))
```

Comment:
The problem is that the macro now returns the doc instead of the generic function.  Use `prog1` or `symbol-function` to return the function.  See `define-class`.